# 010 :  Who owns the plants and their turbines across the USA --> Footprint of turbine owners
* will be joining turbines --> operators table on eia_id --> plant_id (gives us plant identification and plant name)
* place the plants on a map --> geographic distribution/ density --> get plant_state 
* --> get operator_name, operator_id for each plant.  

In [2]:
# import necessary modules
import requests
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


%matplotlib inline

# import .csv data

In [3]:
# import turbines data

turbines = pd.read_csv('wind_turbine_20220114.csv')
turbines

case_id faa_ors faa_asn  usgs_pr_id   eia_id t_state  \
0      3000001     NaN     NaN      6569.0  54686.0      CA   
1      3000003     NaN     NaN      3100.0      NaN      CA   
2      3000008     NaN     NaN      3463.0  10597.0      CA   
3      3000016     NaN     NaN     15543.0  50535.0      CA   
4      3000020     NaN     NaN      1944.0      NaN      AK   
...        ...     ...     ...         ...      ...     ...   
70803  3118667     NaN     NaN         NaN      NaN      NM   
70804  3118668     NaN     NaN         NaN      NaN      NM   
70805  3118669     NaN     NaN         NaN      NaN      NM   
70806  3118670     NaN     NaN         NaN      NaN      NM   
70807  3118671     NaN     NaN         NaN      NaN      NM   

                 t_county  t_fips                                  p_name  \
0             Kern County    6029                               Difwind 6   
1             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
2             Kern County    6029                                Ridgetop   
3        Riverside County    6065                          Mesa Wind Farm   
4      Bethel Census Area    2050                             Tuntutuliak   
...                   ...     ...                                     ...   
70803       Catron County   35003                   unknown Catron County   
70804       Catron County   35003                   unknown Catron County   
70805       Catron County   35003                   unknown Catron County   
70806       Catron County   35003                   unknown Catron County   
70807     Torrance County   35057                 unknown Torrance County   

       p_year  ...   t_rsa  t_ttlh retrofit retrofit_year  t_conf_atr  \
0      1987.0  ...     NaN     NaN        0           NaN           1   
1      1982.0  ...     NaN     NaN        0           NaN           1   
2      1985.0  ...     NaN     NaN        0           NaN           1   
3      1983.0  ...  183.85    30.4        0           NaN           3   
4      2012.0  ...  226.98    39.5        0           NaN           3   
...       ...  ...     ...     ...      ...           ...         ...   
70803     NaN  ...     NaN     NaN        0           NaN           1   
70804     NaN  ...     NaN     NaN        0           NaN           1   
70805     NaN  ...     NaN     NaN        0           NaN           1   
70806     NaN  ...     NaN     NaN        0           NaN           1   
70807     NaN  ...     NaN     NaN        0           NaN           1   

       t_conf_loc  t_img_date     t_img_srce       xlong       ylat  
0               3  2018-11-03  Digital Globe -118.270027  35.066048  
1               3  2018-11-03  Digital Globe -118.299385  35.044796  
2               3  2018-05-08  Digital Globe -118.302780  35.078526  
3               3  2018-11-03  Digital Globe -116.652931  33.954193  
4               3  2020-05-19  Digital Globe -162.661880  60.336800  
...           ...         ...            ...         ...        ...  
70803           3  2021-12-02  Digital Globe -108.901253  34.164696  
70804           3  2021-12-02  Digital Globe -108.907745  34.103058  
70805           3  2021-12-02  Digital Globe -108.915382  34.110283  
70806           3  2021-12-02  Digital Globe -108.869209  34.157585  
70807           3  2021-12-05  Digital Globe -105.590172  34.613602  

[70808 rows x 27 columns]

In [4]:
# import turbines data

operators = pd.read_csv('eia923_operators.csv')
operators

C:\Users\Public\Documents\Wondershare\CreatorTemp\ipykernel_44736\2873213737.py:3: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  operators = pd.read_csv('eia923_operators.csv')


plant_id combined_heat_and_power_plant nuclear_unit_id  \
0             1                             N               .   
1             2                             N               .   
2             3                             N               .   
3             3                             N               .   
4             3                             N               .   
...         ...                           ...             ...   
15110     99999                             N               .   
15111     99999                             N               .   
15112     99999                             N               .   
15113     99999                             N               .   
15114     99999                             N               .   

                       plant_name                   operator_name operator_id  \
0                      Sand Point  TDX Sand Point Generating, LLC       63560   
1                    Bankhead Dam                Alabama Power Co         195   
2                           Barry                Alabama Power Co         195   
3                           Barry                Alabama Power Co         195   
4                           Barry                Alabama Power Co         195   
...                           ...                             ...         ...   
15110  State-Fuel Level Increment      State-Fuel Level Increment       99999   
15111  State-Fuel Level Increment      State-Fuel Level Increment       99999   
15112  State-Fuel Level Increment      State-Fuel Level Increment       99999   
15113  State-Fuel Level Increment      State-Fuel Level Increment       99999   
15114  State-Fuel Level Increment      State-Fuel Level Increment       99999   

      plant_state census_region nerc_region  reserved  ...  netgen_september  \
0              AK          PACN         NaN       NaN  ...              24.0   
1              AL           ESC        SERC       NaN  ...            9187.0   
2              AL           ESC        SERC       NaN  ...          254351.0   
3              AL           ESC        SERC       NaN  ...          486059.0   
4              AL           ESC        SERC       NaN  ...          347191.0   
...           ...           ...         ...       ...  ...               ...   
15110          NV           MTN         NaN       NaN  ...               0.0   
15111          MD           SAT         NaN       NaN  ...               0.0   
15112          NH           NEW         NaN       NaN  ...               0.0   
15113          MA           NEW         NaN       NaN  ...               0.0   
15114          ID           MTN         NaN       NaN  ...               0.0   

       netgen_october netgen_november netgen_december  \
0                29.0            33.0            31.0   
1              9137.0         10514.0         10936.0   
2            220717.0        248068.0        227801.0   
3            420338.0        483657.0        455550.0   
4            424258.0        335570.0        144471.0   
...               ...             ...             ...   
15110             0.0             0.0             0.0   
15111             0.0             0.0             0.0   
15112             0.0             0.0             0.0   
15113             0.0             0.0             0.0   
15114             0.0             0.0             0.0   

      total_fuel_consumption_quantity electric_fuel_consumption_quantity  \
0                                   0                                  0   
1                                   0                                  0   
2                             1219642                            1219642   
3                            52001418                           52001418   
4                             1292499                            1292499   
...                               ...                                ...   
15110                               0                                  0   

# general info on each data frame
* rows and cols
* data type in each col

In [5]:
turbines.shape

(70808, 27)

In [15]:
turbines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70808 entries, 0 to 70807
Data columns (total 27 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   case_id        70808 non-null  int64  
 1   faa_ors        65016 non-null  object 
 2   faa_asn        65686 non-null  object 
 3   usgs_pr_id     38263 non-null  float64
 4   eia_id         65015 non-null  float64
 5   t_state        70808 non-null  object 
 6   t_county       70808 non-null  object 
 7   t_fips         70808 non-null  int64  
 8   p_name         70808 non-null  object 
 9   p_year         70195 non-null  float64
 10  p_tnum         70808 non-null  int64  
 11  p_cap          66326 non-null  float64
 12  t_manu         65168 non-null  object 
 13  t_model        65031 non-null  object 
 14  t_cap          65328 non-null  float64
 15  t_hh           64628 non-null  float64
 16  t_rd           64874 non-null  float64
 17  t_rsa          64874 non-null  float64
 18  t_ttlh

In [6]:
operators.shape

(15115, 97)

In [16]:
operators.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15115 entries, 0 to 15114
Data columns (total 97 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   plant_id                            15115 non-null  int64  
 1   combined_heat_and_power_plant       15115 non-null  object 
 2   nuclear_unit_id                     15115 non-null  object 
 3   plant_name                          15114 non-null  object 
 4   operator_name                       15114 non-null  object 
 5   operator_id                         15115 non-null  object 
 6   plant_state                         15114 non-null  object 
 7   census_region                       15115 non-null  object 
 8   nerc_region                         14550 non-null  object 
 9   reserved                            0 non-null      float64
 10  naics_code                          15115 non-null  int64  
 11  eia_sector_number                   15115

# count eia_id nulls

In [10]:
# count nulls in eia_id
eia_id_nulls = turbines['eia_id'].isnull().sum()
print(eia_id_nulls)

5793


In [9]:
# count values --> string "NULL" in eia_id
# remember conditionals are in parentheses (      )
eia_id_nulls_String = (turbines['eia_id'] == 'NULL').sum()
print(eia_id_nulls_String)

# there are no 'NULL' string values --> 

0


In [14]:
# get percentage of nulls in eia_id

percent_eia_nulls = eia_id_nulls / turbines.shape[0] * 100
percent_eia_nulls

8.1812789515309

8.18% nulls in eia_id

### are projects the same as plants?
* ie is there only 1 plant_id per p_name (project name)
* in turbines : group by p_name --> then do a count of plant_id ==> it should be 1 if they are the same.

*Answer* --> no, not all of the p_names (turbines) have a corresponding plant_name (operators).  The names do not match.   

In [18]:
# count 'unknown' in p_name

p_name_unknown =  turbines['p_name'].str.contains('unknown', case=False).sum()
print(p_name_unknown)

# so there are less p_name unknowns than there are eia_id nulls

# but it seems that even

4535


In [21]:
turbines_unique_p_name = turbines['p_name'].value_counts()
turbines_unique_p_name

# 1708 unique project names

p_name
unknown Tehachapi Wind Resource Area 1        713
Stateline Wind Project                        454
Mesa Wind Farm                                432
unknown Torrance County                       355
Mojave 16, 17 & 18 (Desertwind III PPC Tru    288
                                             ... 
Teichert Aggregates                             1
GobNob                                          1
BTI Wind                                        1
Brookfield Tehachapi 3                          1
unknown Lyon County                             1
Name: count, Length: 1708, dtype: int64

In [23]:
operators_unique_plant_name = operators['plant_name'].value_counts()
operators_unique_plant_name

plant_name
State-Fuel Level Increment       162
Mansfield Mill                    11
Haynes                            11
St Clair                          11
Paradise                          10
                                ... 
WWRF Solar Plant                   1
Panther Creek Wind Farm II         1
Panther Creek Wind Farm I          1
FPL Energy Ashtabula Wind LLC      1
EQX005.0 Toyama Fuel Cell          1
Name: count, Length: 10448, dtype: int64

In [24]:
# create data frame with eia_id null rows removed
turbines_nullRem = turbines.dropna(subset=['eia_id']).copy()
turbines_nullRem

case_id    faa_ors           faa_asn  usgs_pr_id   eia_id t_state  \
0      3000001        NaN               NaN      6569.0  54686.0      CA   
2      3000008        NaN               NaN      3463.0  10597.0      CA   
3      3000016        NaN               NaN     15543.0  50535.0      CA   
5      3000025        NaN               NaN     12159.0  10005.0      CA   
8      3000039        NaN               NaN     15579.0  50535.0      CA   
...        ...        ...               ...         ...      ...     ...   
70626  3117926  40-003040  2020-WTW-1408-OE         NaN  56335.0      OK   
70639  3117954  27-002696  2020-WTE-1999-OE         NaN  56424.0      MN   
70682  3118028  36-020197  2019-WTE-8213-OE         NaN  56575.0      NY   
70760  3118173  04-020391   2020-WTW-850-OE         NaN  58155.0      AZ   
70773  3118363  19-022438               NaN         NaN  57875.0      IA   

               t_county  t_fips          p_name  p_year  ...    t_rsa  t_ttlh  \
0           Kern County    6029       Difwind 6  1987.0  ...      NaN     NaN   
2           Kern County    6029        Ridgetop  1985.0  ...      NaN     NaN   
3      Riverside County    6065  Mesa Wind Farm  1983.0  ...   183.85    30.4   
5         Merced County    6047  Dinosaur Point  1988.0  ...      NaN     NaN   
8      Riverside County    6065  Mesa Wind Farm  1983.0  ...   183.85    30.4   
...                 ...     ...             ...     ...  ...      ...     ...   
70626      Kiowa County   40075  Blue Canyon II  2005.0  ...  5026.55   107.0   
70639      Mower County   27099    Mower County  2006.0  ...  9160.88   134.0   
70682       Erie County   36029     Steel Winds  2007.0  ...  6792.91   126.5   
70760   Coconino County    4005    Perrin Ranch  2012.0  ...  5345.62   121.3   
70773      Adair County   19001   Morning Light  2012.0  ...  9160.88   134.1   

      retrofit retrofit_year  t_conf_atr  t_conf_loc  t_img_date  \
0            0           NaN           1           3  2018-11-03   
2            0           NaN           1           3  2018-05-08   
3            0           NaN           3           3  2018-11-03   
5            0           NaN           1           3  2018-09-19   
8            0           NaN           3           3  2018-11-03   
...        ...           ...         ...         ...         ...   
70626        0           NaN           3           3  2021-10-06   
70639        1        2020.0           2           3  2020-07-31   
70682        0           NaN           3           3  2021-11-24   
70760        0           NaN           3           3  2019-10-20   
70773        0           NaN           2           3  2020-08-18   

          t_img_srce       xlong       ylat  
0      Digital Globe -118.270027  35.066048  
2      Digital Globe -118.302780  35.078526  
3      Digital Globe -116.652931  33.954193  
5      Digital Globe -121.170876  37.035725  
8      Digital Globe -116.658089  33.958694  
...              ...         ...        ...  
70626  Digital Globe  -98.642632  34.884048  
70639  Digital Globe  -92.654350  43.566082  
70682  Digital Globe  -78.864914  42.813736  
70760  Digital Globe -112.296204  35.417610  
70773  Digital Globe  -94.574005  41.401367  

[65015 rows x 27 columns]

In [26]:
# rename column eia_id to plant_id
turbines_nullRem = turbines_nullRem.rename(columns={'eia_id': 'plant_id'})

In [25]:
# create data frame with the removed rows from the step above
# ie rows with a null eia
# we will use p_name --> project name to join these to operators df

turbines_null_eia = turbines[turbines['eia_id'].isnull()].copy()
turbines_null_eia

case_id faa_ors faa_asn  usgs_pr_id  eia_id t_state  \
1      3000003     NaN     NaN      3100.0     NaN      CA   
4      3000020     NaN     NaN      1944.0     NaN      AK   
6      3000026     NaN     NaN      6696.0     NaN      CA   
7      3000034     NaN     NaN      5183.0     NaN      CA   
9      3000053     NaN     NaN      6912.0     NaN      CA   
...        ...     ...     ...         ...     ...     ...   
70803  3118667     NaN     NaN         NaN     NaN      NM   
70804  3118668     NaN     NaN         NaN     NaN      NM   
70805  3118669     NaN     NaN         NaN     NaN      NM   
70806  3118670     NaN     NaN         NaN     NaN      NM   
70807  3118671     NaN     NaN         NaN     NaN      NM   

                 t_county  t_fips                                  p_name  \
1             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
4      Bethel Census Area    2050                             Tuntutuliak   
6             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
7             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
9             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
...                   ...     ...                                     ...   
70803       Catron County   35003                   unknown Catron County   
70804       Catron County   35003                   unknown Catron County   
70805       Catron County   35003                   unknown Catron County   
70806       Catron County   35003                   unknown Catron County   
70807     Torrance County   35057                 unknown Torrance County   

       p_year  ...   t_rsa  t_ttlh retrofit retrofit_year  t_conf_atr  \
1      1982.0  ...     NaN     NaN        0           NaN           1   
4      2012.0  ...  226.98    39.5        0           NaN           3   
6      1982.0  ...     NaN     NaN        0           NaN           1   
7      1982.0  ...     NaN     NaN        0           NaN           1   
9      1982.0  ...     NaN     NaN        0           NaN           1   
...       ...  ...     ...     ...      ...           ...         ...   
70803     NaN  ...     NaN     NaN        0           NaN           1   
70804     NaN  ...     NaN     NaN        0           NaN           1   
70805     NaN  ...     NaN     NaN        0           NaN           1   
70806     NaN  ...     NaN     NaN        0           NaN           1   
70807     NaN  ...     NaN     NaN        0           NaN           1   

       t_conf_loc  t_img_date     t_img_srce       xlong       ylat  
1               3  2018-11-03  Digital Globe -118.299385  35.044796  
4               3  2020-05-19  Digital Globe -162.661880  60.336800  
6               3  2018-11-03  Digital Globe -118.267090  35.072395  
7               3  2018-05-08  Digital Globe -118.370842  35.073112  
9               3  2018-10-28  Digital Globe -118.253685  35.079994  
...           ...         ...            ...         ...        ...  
70803           3  2021-12-02  Digital Globe -108.901253  34.164696  
70804           3  2021-12-02  Digital Globe -108.907745  34.103058  
70805           3  2021-12-02  Digital Globe -108.915382  34.110283  
70806           3  2021-12-02  Digital Globe -108.869209  34.157585  
70807           3  2021-12-05  Digital Globe -105.590172  34.613602  

[5793 rows x 27 columns]

# join .csv data
* join on LHS table:  (turbines) --> plant_id  = plant identification number
* join on RHS table:  (operators) --> plant_id


In [27]:
# rename turbines['eia_id'] --> turbines['plant_id']
turbines = turbines.rename(columns={'eia_id': 'plant_id'})


In [29]:
# do a left join turbines --> operators on 'plant_id'
# this will produce nulls in the plant_id col

turbines_combined = pd.merge(turbines,operators, on='plant_id', how='left')
turbines_combined

case_id faa_ors faa_asn  usgs_pr_id  plant_id t_state  \
0      3000001     NaN     NaN      6569.0   54686.0      CA   
1      3000003     NaN     NaN      3100.0       NaN      CA   
2      3000008     NaN     NaN      3463.0   10597.0      CA   
3      3000016     NaN     NaN     15543.0   50535.0      CA   
4      3000020     NaN     NaN      1944.0       NaN      AK   
...        ...     ...     ...         ...       ...     ...   
71820  3118667     NaN     NaN         NaN       NaN      NM   
71821  3118668     NaN     NaN         NaN       NaN      NM   
71822  3118669     NaN     NaN         NaN       NaN      NM   
71823  3118670     NaN     NaN         NaN       NaN      NM   
71824  3118671     NaN     NaN         NaN       NaN      NM   

                 t_county  t_fips                                  p_name  \
0             Kern County    6029                               Difwind 6   
1             Kern County    6029  unknown Tehachapi Wind Resource Area 1   
2             Kern County    6029                                Ridgetop   
3        Riverside County    6065                          Mesa Wind Farm   
4      Bethel Census Area    2050                             Tuntutuliak   
...                   ...     ...                                     ...   
71820       Catron County   35003                   unknown Catron County   
71821       Catron County   35003                   unknown Catron County   
71822       Catron County   35003                   unknown Catron County   
71823       Catron County   35003                   unknown Catron County   
71824     Torrance County   35057                 unknown Torrance County   

       p_year  ...  netgen_september  netgen_october netgen_november  \
0      1987.0  ...            1441.0          1570.0          1202.0   
1      1982.0  ...               NaN             NaN             NaN   
2      1985.0  ...            2317.0          2525.0          1932.0   
3      1983.0  ...              48.0            52.0            40.0   
4      2012.0  ...               NaN             NaN             NaN   
...       ...  ...               ...             ...             ...   
71820     NaN  ...               NaN             NaN             NaN   
71821     NaN  ...               NaN             NaN             NaN   
71822     NaN  ...               NaN             NaN             NaN   
71823     NaN  ...               NaN             NaN             NaN   
71824     NaN  ...               NaN             NaN             NaN   

      netgen_december  total_fuel_consumption_quantity  \
0              1180.0                              0.0   
1                 NaN                              NaN   
2              1897.0                              0.0   
3                39.0                              0.0   
4                 NaN                              NaN   
...               ...                              ...   
71820             NaN                              NaN   
71821             NaN                              NaN   
71822             NaN                              NaN   
71823             NaN                              NaN   
71824             NaN                              NaN   

       electric_fuel_consumption_quantity  total_fuel_consumption_mm_btu  \
0                                     0.0                       215326.0   
1                                     NaN                            NaN   
2                                     0.0                       346165.0   
3                                     0.0                         7124.0   
4                                     NaN                            NaN   
...                                   ...                            ...   
71820                                 NaN                            NaN   
71821                                 NaN                            NaN   
71822                                 NaN                          

### get percentage of null plant_id in the resultant combined df

In [30]:


# count nulls in plant_id
plant_id_nulls = turbines_combined['plant_id'].isnull().sum()
print(plant_id_nulls)

percent_plant_nulls = plant_id_nulls / turbines_combined.shape[0] * 100
print(percent_plant_nulls)
print(percent_eia_nulls)

5793
8.065436825617821
8.1812789515309


In [31]:
# rename cols
# p_name--> project_name
# t_state --> state
# t_county --> county

turbines_combined = turbines_combined.rename(columns={
    'p_name': 'project_name',
    't_state': 'state',
    't_county': 'county'
})

# drop the redundant cols

In [32]:
selected_columns = ['plant_id', 'project_name', 'plant_name', 'operator_name', 'xlong', 'ylat', 'operator_id', 'state', 'county', 'p_tnum']

turbines_combined = turbines_combined[selected_columns]
turbines_combined

plant_id                            project_name            plant_name  \
0       54686.0                               Difwind 6  Difwind Farms Ltd VI   
1           NaN  unknown Tehachapi Wind Resource Area 1                   NaN   
2       10597.0                                Ridgetop   Ridgetop Energy LLC   
3       50535.0                          Mesa Wind Farm  Mesa Wind Power Corp   
4           NaN                             Tuntutuliak                   NaN   
...         ...                                     ...                   ...   
71820       NaN                   unknown Catron County                   NaN   
71821       NaN                   unknown Catron County                   NaN   
71822       NaN                   unknown Catron County                   NaN   
71823       NaN                   unknown Catron County                   NaN   
71824       NaN                 unknown Torrance County                   NaN   

                                       operator_name       xlong       ylat  \
0                        Terra-Gen Operating Co-Wind -118.270027  35.066048   
1                                                NaN -118.299385  35.044796   
2                        Terra-Gen Operating Co-Wind -118.302780  35.078526   
3      Brookfield Renewable Trading and Marketing LP -116.652931  33.954193   
4                                                NaN -162.661880  60.336800   
...                                              ...         ...        ...   
71820                                            NaN -108.901253  34.164696   
71821                                            NaN -108.907745  34.103058   
71822                                            NaN -108.915382  34.110283   
71823                                            NaN -108.869209  34.157585   
71824                                            NaN -105.590172  34.613602   

      operator_id state              county  p_tnum  
0            2770    CA         Kern County     250  
1             NaN    CA         Kern County     713  
2            2770    CA         Kern County     236  
3           63235    CA    Riverside County     460  
4             NaN    AK  Bethel Census Area       5  
...           ...   ...                 ...     ...  
71820         NaN    NM       Catron County      34  
71821         NaN    NM       Catron County      34  
71822         NaN    NM       Catron County      34  
71823         NaN    NM       Catron County      34  
71824         NaN    NM     Torrance County     355  

[71825 rows x 10 columns]

### reorder cols

In [34]:
column_order = ['plant_id', 'plant_name', 'project_name', 'operator_name', 'operator_id', 'state', 'county', 'xlong', 'ylat', 'p_tnum']
turbines_combined = turbines_combined[column_order]
turbines_combined

plant_id            plant_name                            project_name  \
0       54686.0  Difwind Farms Ltd VI                               Difwind 6   
1           NaN                   NaN  unknown Tehachapi Wind Resource Area 1   
2       10597.0   Ridgetop Energy LLC                                Ridgetop   
3       50535.0  Mesa Wind Power Corp                          Mesa Wind Farm   
4           NaN                   NaN                             Tuntutuliak   
...         ...                   ...                                     ...   
71820       NaN                   NaN                   unknown Catron County   
71821       NaN                   NaN                   unknown Catron County   
71822       NaN                   NaN                   unknown Catron County   
71823       NaN                   NaN                   unknown Catron County   
71824       NaN                   NaN                 unknown Torrance County   

                                       operator_name operator_id state  \
0                        Terra-Gen Operating Co-Wind        2770    CA   
1                                                NaN         NaN    CA   
2                        Terra-Gen Operating Co-Wind        2770    CA   
3      Brookfield Renewable Trading and Marketing LP       63235    CA   
4                                                NaN         NaN    AK   
...                                              ...         ...   ...   
71820                                            NaN         NaN    NM   
71821                                            NaN         NaN    NM   
71822                                            NaN         NaN    NM   
71823                                            NaN         NaN    NM   
71824                                            NaN         NaN    NM   

                   county       xlong       ylat  p_tnum  
0             Kern County -118.270027  35.066048     250  
1             Kern County -118.299385  35.044796     713  
2             Kern County -118.302780  35.078526     236  
3        Riverside County -116.652931  33.954193     460  
4      Bethel Census Area -162.661880  60.336800       5  
...                   ...         ...        ...     ...  
71820       Catron County -108.901253  34.164696      34  
71821       Catron County -108.907745  34.103058      34  
71822       Catron County -108.915382  34.110283      34  
71823       Catron County -108.869209  34.157585      34  
71824     Torrance County -105.590172  34.613602     355  

[71825 rows x 10 columns]

### fill where operator_name is null with 'unknown'

In [35]:
turbines_combined['operator_name'] = turbines_combined['operator_name'].fillna('unknown')
turbines_combined

plant_id            plant_name                            project_name  \
0       54686.0  Difwind Farms Ltd VI                               Difwind 6   
1           NaN                   NaN  unknown Tehachapi Wind Resource Area 1   
2       10597.0   Ridgetop Energy LLC                                Ridgetop   
3       50535.0  Mesa Wind Power Corp                          Mesa Wind Farm   
4           NaN                   NaN                             Tuntutuliak   
...         ...                   ...                                     ...   
71820       NaN                   NaN                   unknown Catron County   
71821       NaN                   NaN                   unknown Catron County   
71822       NaN                   NaN                   unknown Catron County   
71823       NaN                   NaN                   unknown Catron County   
71824       NaN                   NaN                 unknown Torrance County   

                                       operator_name operator_id state  \
0                        Terra-Gen Operating Co-Wind        2770    CA   
1                                            unknown         NaN    CA   
2                        Terra-Gen Operating Co-Wind        2770    CA   
3      Brookfield Renewable Trading and Marketing LP       63235    CA   
4                                            unknown         NaN    AK   
...                                              ...         ...   ...   
71820                                        unknown         NaN    NM   
71821                                        unknown         NaN    NM   
71822                                        unknown         NaN    NM   
71823                                        unknown         NaN    NM   
71824                                        unknown         NaN    NM   

                   county       xlong       ylat  p_tnum  
0             Kern County -118.270027  35.066048     250  
1             Kern County -118.299385  35.044796     713  
2             Kern County -118.302780  35.078526     236  
3        Riverside County -116.652931  33.954193     460  
4      Bethel Census Area -162.661880  60.336800       5  
...                   ...         ...        ...     ...  
71820       Catron County -108.901253  34.164696      34  
71821       Catron County -108.907745  34.103058      34  
71822       Catron County -108.915382  34.110283      34  
71823       Catron County -108.869209  34.157585      34  
71824     Torrance County -105.590172  34.613602     355  

[71825 rows x 10 columns]

# write to csv

In [37]:
turbines_combined.to_csv('turbines_data_cleaned.csv', index=False)

### save all the required modules that I have been using for this project

pip freeze > requirements.txt

In [38]:
pip freeze > requirements.txt

Note: you may need to restart the kernel to use updated packages.
